drivers: sensor: Initial commit of TDK Invensense ICM42605

JHKim8212 opened the pull request:
drivers: sensor: Initial driver and sample app of TDK Invensense ICM42605

Initial driver and sample application of  TDK Invensense ICM42605 6-axis motion sensor.
This driver provide DTS for nRF52 DK board DTS setting.

Providing features are below.
1. Sensor data streaming - Accel, gyro
2. Tap, Double tap triggering.
3. Set/Get FSR, ODR by set attr API

Signed-off-by : JuHyun Kim <jkim@invensense.com>

JHKim8212 commented:
Hello.

I got compliance failure while commit.
below is log.
This driver using KCONFIG symbol in sample/sensor/icm42605 sample application (prj.conf contain below symbols.)
What should I fix for below error?


##[error]See https://docs.zephyrproject.org/latest/guides/kconfig/index.html for more details.

Found references to undefined Kconfig symbols. If any of these are false
positives, then add them to UNDEF_KCONFIG_WHITELIST in check_compliance.py in the ci-tools repo.

If the reference is for a comment like /* CONFIG_FOO_* */ (or
/* CONFIG_FOO_*_... */), then please use exactly that form (with the '*'). The
CI check knows not to flag it.

More generally, a reference followed by $, @, {, *, or ## will never be
flagged.

CONFIG_ICM42605                            drivers/sensor/icm42605/CMakeLists.txt:6, drivers/sensor/icm42605/CMakeLists.txt:7, samples/sensor/icm42605/prj.conf:9
CONFIG_ICM42605_THREAD_PRIORITY            drivers/sensor/icm42605/icm42605_trigger.c:134
CONFIG_ICM42605_THREAD_STACK_SIZE          drivers/sensor/icm42605/icm42605.h:54, drivers/sensor/icm42605/icm42605_trigger.c:132
CONFIG_ICM42605_TRIGGER                    drivers/sensor/icm42605/CMakeLists.txt:9, samples/sensor/icm42605/src/main.c:89, samples/sensor/icm42605/src/main.c:103
CONFIG_ICM42605_TRIGGER_OWN_THREAD         samples/sensor/icm42605/prj.conf:10
##[error]Process completed with exit code 1.

gmarull commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,392 @@
+{
+	struct icm42605_data *drv_data = dev->data;
+
+	if (val == NULL) {
```
An assertion would be better, as having it NULL is a programming error.

JHKim8212 replied:
Good point. I will fix and will test.

JHKim8212 replied:
Fixed

gmarull commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,392 @@
+			if (val->val1 > 1000) {
+				drv_data->accel_hz = 1000;
+			} else if (val->val1 < 12) {
+				drv_data->accel_hz = 12;
```
These constants should be defined and/or documented

gmarull commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,392 @@
+	.int_flags = DT_INST_GPIO_FLAGS(0, int_gpios),
+};
+
+DEVICE_AND_API_INIT(icm42605, DT_INST_LABEL(0),
```
Now that it's possible I think it's better to create multi-instance drivers. You can search for `DT_INST_FOREACH_STATUS_OKAY` e.g. https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/kscan/kscan_ft5336.c#L275

JHKim8212 replied:
I will implement this feature at next time

JHKim8212 replied:
OK I already implemented and tested, I will commit soon.

JHKim8212 replied:
@gmarull Can you please review latest implementation?
I followed multi instance driver like link what you shared.

gmarull commented on drivers/sensor/icm42605/icm42605_setup.c:
```diff
@@ -0,0 +1,377 @@
+	LOG_DBG("Read REG_INTF_CONFIG0 0x%X", v);
+
+	/* 0x03 -> Disable i2c */
+	v |= 0x03;
```
This setting should be defined

JHKim8212 replied:
OK

gmarull commented on drivers/sensor/icm42605/icm42605_setup.c:
```diff
@@ -0,0 +1,377 @@
+	if (r) {
+		return r;
+	}
+	v = 0;
```
blank line between operations would improve readability

JHKim8212 replied:
OK

gmarull commented on drivers/sensor/icm42605/icm42605_setup.c:
```diff
@@ -0,0 +1,377 @@
+	int1_en = 0;
+	smd = 0;
+
+	int0_en = BIT_INT_UI_DRDY_INT1_EN;
```
should be set after fifo config 1

JHKim8212 replied:
Fixed

gmarull commented on drivers/sensor/icm42605/icm42605_setup.c:
```diff
@@ -0,0 +1,377 @@
+
+	if (drv_data->tap_en) {
+		v = BIT_TAP_ENABLE;
+		r = inv_spi_single_write(REG_APEX_CONFIG0, &v);
```
error is not checked here?

gmarull replied:
another comment: would be better to be consistent on either using `r` or `result`

JHKim8212 replied:
Fixed

gmarull commented on drivers/sensor/icm42605/icm42605_setup.c:
```diff
@@ -0,0 +1,377 @@
+
+	result = inv_spi_single_write(REG_PWR_MGMT0, &v);
+
+	k_msleep(200);
```
this sleep time should be documented

gmarull commented on drivers/sensor/icm42605/icm42605_setup.c:
```diff
@@ -0,0 +1,377 @@
+
+	if (drv_data->sensor_started) {
+		LOG_ERR("Sensor already started");
+		return 0;
```
returning 0 on error?

JHKim8212 replied:
Fixed

gmarull commented on drivers/sensor/icm42605/icm42605_spi.c:
```diff
@@ -0,0 +1,99 @@
+
+	if (err) {
+		LOG_ERR("SPI error: %d", err);
+		return -EIO;
```
better return `r` so that error code is propagated.

gmarull commented on drivers/sensor/icm42605/icm42605_spi.c:
```diff
@@ -0,0 +1,99 @@
+	err = spi_write(spi_dev, spi_cfg, &tx);
+
+	if (err) {
+		LOG_ERR("SPI error: %d", err);
```
This log may be redundant if SPI logging is enabled.

gmarull commented on drivers/sensor/icm42605/icm42605_spi.c:
```diff
@@ -0,0 +1,99 @@
+{
+	spi_dev = spi_device;
+
+	if (!spi_dev) {
```
An assertion would be better, or even nothing.

gmarull commented on drivers/sensor/icm42605/icm42605_spi.c:
```diff
@@ -0,0 +1,99 @@
+	return 0;
+}
+
+int icm42605_spi_init(struct device *spi_device, struct spi_config *spi_config)
```
What is the purpose of this function?

JHKim8212 replied:
When I tried to build icm42605_spi.c file with using DT_INST_~~~~ APIs, it generate build error.
SPI read and write API is used icm42605.c and icm42605_setup.c , so I seperated spi_read and spi_write to other c file.
If there is any easy or correct way to set SPI instance out from icm42605.c file, I will use that way.

gmarull commented on drivers/sensor/icm42605/icm42605_trigger.c:
```diff
@@ -0,0 +1,138 @@
+	gpio_pin_interrupt_configure(drv_data->gpio, cfg->int_pin,
+				     GPIO_INT_EDGE_TO_ACTIVE);
+
+	k_msleep(100);
```
Sleep time should be defined and documented.

JHKim8212 replied:
I will document soon.

gmarull commented on drivers/sensor/icm42605/icm42605_trigger.c:
```diff
@@ -0,0 +1,138 @@
+
+	if (gpio_add_callback(drv_data->gpio, &drv_data->gpio_cb) < 0) {
+		LOG_ERR("Failed to set gpio callback");
+		return -EIO;
```
better propagate error code

gmarull commented on drivers/sensor/icm42605/icm42605_trigger.c:
```diff
@@ -0,0 +1,138 @@
+	if (drv_data->gpio == NULL) {
+		LOG_ERR("Failed to get pointer to %s device",
+			cfg->int_label);
+		return -EINVAL;
```
`-ENODEV`

JHKim8212 replied:
Fixed

gmarull commented on samples/sensor/icm42605/src/main.c:
```diff
@@ -0,0 +1,126 @@
+	ACCEL_FS_8G,
+	ACCEL_FS_4G,
+	ACCEL_FS_2G,
+};
```
These enumerations should be provided, e.g. in a public header here https://github.com/zephyrproject-rtos/zephyr/tree/master/include/drivers/sensor

gmarull replied:
Why is sample removed?

JHKim8212 replied:
Current pull request is related with Sensor Driver.
So I removed sample application.
I will commit sample application after this driver is merged.

gmarull replied:
It's ok to add sample application here, as soon as it has its own commit.

JHKim8212 replied:
OK. I will add sample application with new commit soon.
(Also add some commit and will squash. :) )

JHKim8212 replied:
Committed sample application.

gmarull commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,392 @@
+		if (data[0] & BIT_INT_STATUS_TAP_DET) {
+			LOG_DBG("Tap detected!!!");
+			result = inv_spi_read(REG_APEX_DATA4, data, 1);
+			if (data[0] & 0x08) {
```
constants should be defined. Another question here, what does driver do in case of tap detection, only logging?

JHKim8212 replied:
I misses value set of tap and double tap, I will implement that feature.

JHKim8212 replied:
Implemented trigger of tap/double tap.
Sample application will be commit at next time.

gmarull commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,392 @@
+				      drv_data->gyro_sensitivity_x10);
+		break;
+	default: /* chan == SENSOR_CHAN_DIE_TEMP */
+		icm42605_convert_temp(val, drv_data->temp);
```
default case should return `-ENOTSUP`.

JHKim8212 replied:
OK Will change.

JHKim8212 replied:
Fixed

gmarull commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,392 @@
+{
+	int result = 0;
+	uint16_t fifo_count = 0;
+	static uint8_t data[HARDWARE_FIFO_SIZE] = { 0, };
```
Why is this variable static? If static should be part of the driver instance.

JHKim8212 replied:
I will remove static
data variable doesn't global.

JHKim8212 replied:
Due to stack limitation, I changed static to driver instance.

gmarull reviewed (changes requested):
I've done an initial review, so I have probably missed things. Overall looks good, I added a few suggestions to improve the driver and I have left some questions to clarify some areas of the code.

Regarding commits, as @carlescufi said you should squash fixes. Another suggestion is to just use lowercase names, e.g. `Driver: Sensor:` -> `drivers: sensor:`

JHKim8212 commented on drivers/sensor/icm42605/Kconfig:
```diff
@@ -0,0 +1,69 @@
+
+config ICM42605_GYRO_FS
+	int "Gyroscope full-scale range"
+	default 2000
```
Thank you so much.
This is my first commit of Zephyr driver and also first experience of Zephyr project, So I confused many things..
I will check links what you provides and will fix soon.

gmarull replied:
Don't hesitate to ask questions if you have doubts. You can also join Zephyr Slack channel.

JHKim8212 replied:
Can you please give me link of Zephyr slack channel?
I will join and will ask if have questions.

JHKim8212 replied:
configure value of FS are not requested for this sensor. So I removed configure of FS

JHKim8212 commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,459 @@
+					      drv_data->fifo_data, 1);
+			if (drv_data->fifo_data[0] & APEX_TAP) {
+				if (drv_data->tap_handler) {
+					if (drv_data->tap_trigger.type ==
```
Do you means ( if ((drv_data->fifo_data[0] & APEX_TAP) && drv_data->tap_handler && drv_data->tap_trigger.type == SENSOR_TRIG_TAP) ?? or just check if (drv_data->tap_handler && drv_data->tap_trigger.type == SENSOPR_TRIG_TAP) ?

gmarull replied:
If tap handler can be null then the first check, otherwise the second is sufficient

JHKim8212 replied:
Fixed

JHKim8212 commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,459 @@
+
+				LOG_DBG("Double Tap");
+			} else {
+				LOG_DBG("Not supported");
```
I will add more log and will change more clear message. 
I also think that there are no log message when no handler and not matched trigger type.

JHKim8212 commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,459 @@
+				drv_data->temp =
+					(int16_t)(drv_data->fifo_data[7]);
+			} else {
+				if (!(drv_data->fifo_data[7] ==
```
The reason why I separated if else {if] is compare data is different between fifo_data[0] and fifo_data[7].
if same data compare, combine is right, but I think that different compare data should separated other else  and if statement.
If it is incorrect, I will combine that.
(If combined else if is more readable, I will change.)

gmarull replied:
I'm fine with your preference here

MaureenHelm replied:
It looks like you don't need `fifo_data` outside the scope of this function. If that's right, you could read the fifo data into a function local variable, which would simplify these byte manipulation expressions quite a bit. I also don't really see the point in masking byte-sized quantities with 0xff.

JHKim8212 replied:
@MaureenHelm When I tested local variable of fifo_data, sensor sample application is stuck.
(Look like stack allocation failure, but cannot debugging this issue on my Nordic nRF52 board)
That is why I moved fifo_data to global, and fifo_data variable also used for icm42605_tap_fetch function.
For prevent frequent stack allocation of fifo_data variable, I think using global variable is more good.

gmarull reviewed (commented):
I think the main remaining thing that would be good to have is the addition of accelerometre/gyroscope initial scale set via DT properties. You can use two device tree entries with an enum on each for that. Other comments are just nits. Remember to squash fix commits, too.

JHKim8212 commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,464 @@
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_ACCEL_XYZ:
+		if (attr == SENSOR_ATTR_SAMPLING_FREQUENCY) {
+			if (val->val1 > 1000) {
```
For SAMPLEING_FREQUENCY, I think current implementation doesn't make confusion even incorrect input.
If val1 is 1001, accel_hz will be 1000 and if val1 is 11 or -100 or any minus value, it will be 12.
(val1 1000 or 12 will be correct value to accel_hz)
And Sampling frequency limitation is from 1000Hz to 12Hz.

JHKim8212 commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,464 @@
+	data->gyro_hz = 10;
+
+	data->accel_sf = ACCEL_FS_16G;
+	data->gyro_sf = GYRO_FS_2000DPS;
```
@gmarull Implemented DT properties based accel/gyro default value initializing feature.
Please review :)

gmarull commented on dts/bindings/sensor/invensense,icm42605.yaml:
```diff
@@ -0,0 +1,61 @@
+        Set to indicate clock leading edge is falling (CPOL=1).
+        For this driver using this property requires also using cpha.
+
+    accel-default-hz:
```
You can name this field `accel-hz` and use the `default` property to assign a default, so that DT entries not specifying this option will get the device default. According to the datasheet Hz/FS values can take limited values, so the driver should reflect that. In case of DT, adding an `enum` with accepted values is what is usually done.

![image](https://user-images.githubusercontent.com/25011557/95179866-82d4d180-07c1-11eb-893e-78e12b3a5ef3.png)

I think that giving freedom on these values may lead to confusions on the capabilities of the device.

JHKim8212 replied:
I cannot tested 2K to 8K Hz on Nordic board due to hardware limitation.
Anyway, your suggestion is correct and I will modify after test with Logic analyzer.

JHKim8212 replied:
@gmarull Tested From 1K to 8K, without result printing.
With result printing, 1K is maximum frequency for accel and gyro

gmarull commented on drivers/sensor/icm42605/icm42605_setup.c:
```diff
@@ -0,0 +1,427 @@
+
+	databuf &= ~BIT_ACCEL_ODR;
+
+	if (a_rate > 500) {
```
only exact supported values should be handled.

gmarull reviewed (commented):
Some last comments on Hz/FS. You should also combine commits, e.g. one for the driver implementation and a second one for the sample application.

gmarull commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,467 @@
+				drv_data->gyro_hz = val->val1;
+			}
+		} else if (attr == SENSOR_ATTR_FULL_SCALE) {
+			if (val->val1 < GYRO_FS_2000DPS ||
```
Attributes are passed and returned with units (see https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/sensor.h#L266). So for example, for sampling frequency you could get "3.125" (val1 = 3, val2 = 125000). So using a chain of if/else if/else you should map that into your internal enum/definition. Only supported values should be handled.

JHKim8212 replied:
@gmarull I think floating point based frequency make confusion to user.
That is why Invensense doesn't use floating point value for Frequency even current sensor provide floating point value at low frequency.
If Zephyer sensor driver should use floating point based frequency, I will implement.

JHKim8212 replied:
@gmarull Misunderstand.. I will fix it.

JHKim8212 replied:
@gmarull Is there any way to get float value from devicetree?
For use floating point value of HZ and FS, need to get float value from devicetree.
Or I need to separate accel_hz or accel_fs to accel_hz_int, accel_hz_frac / accel_fs_int, accel_fs_frac
Need any idea about this implementation..

                .accel_hz = DT_INST_PROP(index, accel_hz),              \
                .gyro_hz = DT_INST_PROP(index, gyro_hz),                \
                .accel_fs = DT_ENUM_IDX(DT_DRV_INST(index), accel_fs),  \
                .gyro_fs = DT_ENUM_IDX(DT_DRV_INST(index), gyro_fs),    \

gmarull replied:
No, floats are not supported on DT. Solutions I can think of is either use another unit scale (for example if it was voltage mV instead of V), or, use string based enums that then translate to appropriate values.

gmarull replied:
@JHKim8212 Floats are not used by the sensor API anyway, the sensor_value struct is used where you have the integer part and the decimal part both as integers.

JHKim8212 replied:
@gmarull Can sensor driver use enum value for sensor frequency instead of actual value with int and fractional?
Code is too heavy like below..


        if (a_rate.val1 >= 4000) {
                if (a_rate.val1 == 4000 && a_rate.val2 <= 0) {
                        databuf |= BIT_ACCEL_ODR_4000;
                } else {
                        databuf |= BIT_ACCEL_ODR_8000;
                }
        } else if (a_rate.val1 >= 2000) {
                if (a_rate.val1 == 2000 && a_rate.val2 <= 0) {
                        databuf |= BIT_ACCEL_ODR_2000;
                } else {
                        databuf |= BIT_ACCEL_ODR_4000;
                }
        } else if (a_rate.val1 >= 1000) {
                if (a_rate.val1 == 1000 && a_rate.val2 <= 0) {
                        databuf |= BIT_ACCEL_ODR_1000;
                } else {
                        databuf |= BIT_ACCEL_ODR_2000;
                }

JHKim8212 replied:
@gmarull Can you please give me opinion about change actual value of Hz to ENUM?

gmarull commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,467 @@
+	case SENSOR_CHAN_GYRO_Z:
+	case SENSOR_CHAN_GYRO_XYZ:
+		if (attr == SENSOR_ATTR_SAMPLING_FREQUENCY) {
+			if (val->val1 > 1000 || val->val1 < 12) {
```
Non-supported values may still get in without returning error.

gmarull commented on dts/bindings/sensor/invensense,icm42605.yaml:
```diff
@@ -0,0 +1,113 @@
+        6 - +-31.25DPS
+        7 - +-15.625DPS
+      enum:
+        - 0
```
Note that you can also use a string based enum, e.g.:

```suggestion
        - "2000"
```

Then use the `DT_ENUM_IDX(DT_DRV_INST(i), gyro_fs)` macro to retrieve the enum number. https://github.com/zephyrproject-rtos/zephyr/blob/master/include/devicetree.h#L499

JHKim8212 replied:
@gmarull Thank you for this nice suggestion. 
Implemented and tested. works correctly for fullscale of accel and gyro

MaureenHelm commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,467 @@
+}
+
+
+int icm42605_init(struct device *dev)
```
```suggestion
static int icm42605_init(struct device *dev)
```

MaureenHelm commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,467 @@
+	struct icm42605_data *drv_data = dev->data;
+	const struct icm42605_config *cfg = dev->config;
+
+	drv_data->spi = device_get_binding(cfg->spi_label);
+
+	spi_cs.gpio_dev = device_get_binding(cfg->gpio_label);
```
please check if the device pointer is NULL

MaureenHelm commented on drivers/sensor/icm42605/icm42605_setup.c:
```diff
@@ -0,0 +1,449 @@
+
+LOG_MODULE_REGISTER(ICM42605, CONFIG_SENSOR_LOG_LEVEL);
+
+int icm42605_set_fs(struct device *dev, uint16_t a_sf, uint16_t g_sf)
```
```suggestion
int icm42605_set_fs(const struct device *dev, uint16_t a_sf, uint16_t g_sf)
```

MaureenHelm commented on dts/bindings/sensor/invensense,icm42605.yaml:
```diff
@@ -0,0 +1,104 @@
+include: spi-device.yaml
+
+properties:
+    spi-max-frequency:
+      type: int
+      required: true
+      description: Maximum clock frequency of device's SPI interface in Hz
```
This is redundant after including `spi-device.yaml`. Please remove

MaureenHelm reviewed (changes requested):
Thanks for the contribution! As @gmarull said, please squash the fixup commits. You should only need a total of two commits for this PR - one for the driver and one for the sample.

JHKim8212 commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,467 @@
+static inline void icm42605_convert_temp(struct sensor_value *val,
+					 int16_t raw_val)
+{
+	val->val1 = (((int64_t)raw_val * 100) / 207) + 25;
+	val->val2 = ((((int64_t)raw_val * 100) % 207) * 1000000) / 207;
```
@MaureenHelm Yes it wrote in datasheet as a formula like below
Temperature in Degrees Centigrade = (FIFO_TEMP_DATA / 2.07) + 25

JHKim8212 commented on drivers/sensor/icm42605/icm42605.c:
```diff
@@ -0,0 +1,467 @@
+		      SPI_MODE_CPHA | SPI_WORD_SET(8) | SPI_LINES_SINGLE |
+		      SPI_TRANSFER_MSB),
+	.cs = NULL,
+};
```
@MaureenHelm Moved spi_cs and spi_cfg to driver data and works well.
Thank you.

JHKim8212 commented:
@MaureenHelm @gmarull Can you please review?

MaureenHelm commented:
> Thanks for the contribution! As @gmarull said, please squash the fixup commits. You should only need a total of two commits for this PR - one for the driver and one for the sample

There are now two commits, but they both touch the driver and the sample. All of the driver changes should be in one commit, and all of the sample changes should be in the second commit. Please review the [Contribution Workflow](https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow).

Commit:
drivers: sensor: Initial driver of TDK Invensense ICM42605

Initial driver of TDK Invensense ICM42605 6-axis motion sensor.
This driver provide DTS for nRF52 DK board DTS setting.

Providing features are below.

Sensor data streaming - Accel, gyro
Tap, Double tap triggering.
Set/Get FSR, ODR by set attr API
Support multi instance feature

Signed-off-by: JuHyun Kim <jkim@invensense.com>

galak commented on dts/bindings/sensor/invensense,icm42605.yaml:
```diff
@@ -0,0 +1,99 @@
+    accel-hz:
+      type: int
+      required: false
+      default: 12
```
Is 12 the POR default for this in the hardware?

JHKim8212 replied:
Default is 1KHz (1000), but reason of hardware limitation of nRF52 board, cannot test 1KHz sensor output.
(Sensor works correctly 1KHz, but problem while printing log).
So I changed to 12 Hz

galak commented on dts/bindings/sensor/invensense,icm42605.yaml:
```diff
@@ -0,0 +1,99 @@
+    gyro-hz:
+      type: int
+      required: false
+      default: 12
```
Is 12 the POR default for this in the hardware?

JHKim8212 replied:
Default is 1KHz (1000), but reason of hardware limitation of nRF52 board, cannot test 1KHz sensor output.
(Sensor works correctly 1KHz, but problem while printing log).
So I changed to 12 Hz

galak commented on dts/bindings/sensor/invensense,icm42605.yaml:
```diff
@@ -0,0 +1,99 @@
+
+    accel-fs:
+      type: string
+      required: false
```
Any reason this is a string?

JHKim8212 replied:
FS is more good to use string than integer due to it is gravity such like 16g or 8g

galak commented on dts/bindings/sensor/invensense,icm42605.yaml:
```diff
@@ -0,0 +1,99 @@
+    gyro-fs:
+      type: string
+      required: false
+      default: "2000"
```
Is 2000 the POR default for this in the hardware?

JHKim8212 replied:
Yes it is default FS of Gyro sensor.

galak reviewed (changes requested):
* Some DTS questions
* Please also add support to `tests/drivers/build_all` for the sensor.

JHKim8212 commented:
> * Some DTS questions
> * Please also add support to `tests/drivers/build_all` for the sensor.

@galak Is there any guide for that?
What should I do for 'tests/drivers/build_all'

galak commented:
> @galak Is there any guide for that?
> What should I do for 'tests/drivers/build_all'

No particular guide, you'll need to add to `tests/drivers/build_all/sensors_i_z.conf` and whatever dts nodes are needed.

You can look at a commit like the following to see an example:

commit 41da0769c19174dbf603e2f9e027b126a446a9b3
Author: Greg Leach <greg.leach@lairdconnect.com>
Date:   Thu Jul 16 19:07:15 2020 +0100

    sensor: Add Silabs SI7055 support

JHKim8212 commented:
@galak 

> > @galak Is there any guide for that?
> > What should I do for 'tests/drivers/build_all'
> 
> No particular guide, you'll need to add to `tests/drivers/build_all/sensors_i_z.conf` and whatever dts nodes are needed.
> 
> You can look at a commit like the following to see an example:
> 
> commit [41da076](https://github.com/zephyrproject-rtos/zephyr/commit/41da0769c19174dbf603e2f9e027b126a446a9b3)
> Author: Greg Leach [greg.leach@lairdconnect.com](mailto:greg.leach@lairdconnect.com)
> Date: Thu Jul 16 19:07:15 2020 +0100
> 
> ```
> sensor: Add Silabs SI7055 support
> ```

@galak Please verify below.
1. sensor_i_z means sensor configuration of name started i to z.. ?
2. Add below to sensor_i_z.conf for build test.
CONFIG_ICM42605=y
CONFIG_ICM42605_TRIGGER_OWN_THREAD=y

But what I concern is, I already added CONFIG_ICM42605=y and CONFIG_ICM42605_TRIGGER_OWN_THREAD=y in proj.conf of icm42605 sample application
Is there any conflict while build test?

galak commented:
> @galak Please verify below.
> 
> 1. sensor_i_z means sensor configuration of name started i to z.. ?
> 2. Add below to sensor_i_z.conf for build test.
>    CONFIG_ICM42605=y
>    CONFIG_ICM42605_TRIGGER_OWN_THREAD=y
> 
> But what I concern is, I already added CONFIG_ICM42605=y and CONFIG_ICM42605_TRIGGER_OWN_THREAD=y in proj.conf of icm42605 sample application
> Is there any conflict while build test?

Having coverage in the build test is useful since its provides a single place to make sure a change that impacts multiple drivers can easily be checked out.

JHKim8212 commented:
> > @galak Please verify below.
> > 
> > 1. sensor_i_z means sensor configuration of name started i to z.. ?
> > 2. Add below to sensor_i_z.conf for build test.
> >    CONFIG_ICM42605=y
> >    CONFIG_ICM42605_TRIGGER_OWN_THREAD=y
> > 
> > But what I concern is, I already added CONFIG_ICM42605=y and CONFIG_ICM42605_TRIGGER_OWN_THREAD=y in proj.conf of icm42605 sample application
> > Is there any conflict while build test?
> 
> Having coverage in the build test is useful since its provides a single place to make sure a change that impacts multiple drivers can easily be checked out.

@galak OK I will add CONFIG_ICM42605=y and CONFIG_ICM42605_TRIGGER_OWN_THREAD=y . And will do build test.

Commit:
samples: sensor: sample app of ICM42605 sensor

Sample application of ICM42605 6 axis motion sensor

Signed-off-by: JuHyun Kim <jkim@invensense.com>

JHKim8212 commented:
@MaureenHelm commits rebase done. please review.....

JHKim8212 commented:
Close pull request due to not feedback. 
I will re-create new pull request.

MaureenHelm commented:
> Close pull request due to not feedback.
> I will re-create new pull request.

For the future, please don't do this because we lose the review history. Instead, you can ping reviewers on Slack or add the `dev-review` label to bring more attention to your PR.
